feat(kmodal, kprompt): reskin components (#1972)
Kong/kongponents: docs/components/modal.md

# Modal

KModal is a modal window component with overlay.

:::tip NOTE

> Consider using [KPrompt component](/components/prompt) instead if your use case is one of the below:
> * you need a simple pop-up element to ask user to confirm their action
> * you need user's input
> * you don't need much customization options for modal window layout

Reviewer: ```suggestion
If your use-case matches one of the scenarios below, consider using [KPrompt](/components/prompt) instead:
* you need a simple pop-up dialog to ask user to confirm their action
* the user performed an action and you first require them to provide input text
* you don't need many customization options for a simple modal with action buttons
```

  title="Modal"
  @cancel="closeAllModals"
  @proceed="closeAllModals"
/>

```html
<KModal
  action-button-text="Ok"
  :visible="modalVisible"
  title="Modal"
  @cancel="handleModalClose"
  @proceed="handleModalProceed"
/>
```

### actionButtonAppearance

Appearance of action button. See [KButton `appearance`](/components/button#appearance) prop for more details. Defaults to `primary`.

<KButton @click="modal5Visible = true">Modal</KButton>
<KModal
  action-button-appearance="danger"
  :visible="modal5Visible"
  title="Modal"
  @cancel="closeAllModals"
  @proceed="closeAllModals"
/>

```html
<KModal
  action-button-appearance="danger"
  :visible="modalVisible"
  title="Modal"
  @cancel="handleModalClose"
  @proceed="handleModalProceed"
/>
```

### actionButtonDisabled

By default the action button is enabled, but you can pass `true` to this prop should you want to change that.

<KComponent
  v-slot="{ data }"
  :data="{ actionEnabled: true }"
>
  <div class="vertical-container">
    <div>
      <KButton @click="modal6Visible = true">Modal</KButton>
    </div>
    <KInputSwitch
      v-model="data.actionEnabled"
      label="Action button enabled"
    />
  </div>

  <KModal

>     :action-button-disabled="!data.actionEnabled"

Reviewer: since we're talking about disabling the button, I think it should be disabled by default in the example. Applies to **all** disabled examples in the other props too

  title="Modal"
  @cancel="handleModalClose"
  @proceed="handleModalProceed"
/>
```

### hideCancelButton

Use this prop to hide the cancel button. Defaults to `false`.

<KButton @click="modal10Visible = true">Modal</KButton>
<KModal
  hide-cancel-button
  :visible="modal10Visible"
  title="Modal"
  @cancel="closeAllModals"
  @proceed="closeAllModals"
/>

```html
<KModal
  hide-cancel-button
  :visible="modalVisible"
  title="Modal"
  @cancel="handleModalClose"
  @proceed="handleModalProceed"
/>
```

### hideCloseIcon

Prop that allows you to hide close icon next to the title. Defaults to `false`. When no title is passed and `hideCloseIcon` is `true`, the modal header section will be hidden.

<KButton @click="modal11Visible = true">Modal</KButton>
<KModal
  hide-close-icon
  :visible="modal11Visible"
  @cancel="closeAllModals"
  @proceed="closeAllModals"
>
  Modal header section is omitted when no header content is present.
</KModal>

```html
<KModal
  hide-close-icon
  :visible="modalVisible"
  @cancel="handleModalClose"
  @proceed="handleModalProceed"
>
  Modal header section is omitted when no header content is present.
</KModal>
```

### proceedOnEnter

> Whether pressing on Enter should result in KModal emitting the [`proceed` event](#proceed). Defaults to `true`.

Reviewer: This example does not function properly on the docs site; modal stays open and nothing happens.

Reviewer: I think it's because the element isn't focused when opened - this is different behavior than existed previously, so it's broken

Reviewer: I'm not sure this makes sense -- hitting "Enter" on the modal/prompt doesn't do anything regardless. Did this break prior to all of this alpha work?

@kaiarrowood I think you originally implemented this?

Author: It's working - you gotta blur from the KButton: https://www.loom.com/share/d8bc979195c342fb9ce50011b0a751a4?sid=4bd23e94-0cd9-4319-9c25-f2e8ebbba403

Reviewer: This didn't used to be the case. IIRC this prop was added because at one point when the dialog opened you could immediately press enter; otherwise, the prop doesn't make sense..?

Author: I think it works as expected. The focus is set on the button, not on KModal. When I don't set focus on anything - it closes on Enter: https://www.loom.com/share/8c340b2c79524eccaf5c28cb40165fa0?sid=8ce509ca-61dc-4b98-b0e0-e60e7b017feb

Reviewer: What I'm saying is that the modal/prompt should be automatically focused, or an element within it, when opened. This used to work this way, so there has been a regression at some point.

Author: Fixed now (both in sandbox and in docs)

Reviewer 2: Chiming in on expected behavior. As far as focus trap goes, I don't think we ever had it working that `Enter` would immediately do something (I think that's an accessibility no-no, because we would manually be grabbing the user's focus and dropping it wherever we chose which could be confusing for non-sighted) - the only place I did see this behavior was v7 KModalFullscreen (didn't check v8). 

KPrompt/KModal have never supported immediate close with `Enter` and honestly not sure they should. We SHOULD allow immediate close with `Esc` though, that doesn't require any focus change - it's just treated as a native keyboard action that always tries to exit out of whatever is focused.

Reviewer: So then what is the purpose of the `preventProceedOnEnter` prop? How did it ever function?

Reviewer 2: `preventProceedOnEnter` is breaking the tie between pressing Enter key and sending the proceed event when focused on the dialog. I believe the thinking was something like, if the user completely controls the buttons by using the footer slot, we don't want to send a proceed event since we don't know the state of the buttons (ie. you wouldn't send `@proceed` if the button is disabled).

Reviewer: ...then the prop can also be removed

Author: I'm fine with this approach. If needed, host app can take care of performing an action on enter:
```
// inside modal body
<KInput @enter.prevent="...
```

Author: Removed the prop

<KComponent
  v-slot="{ data }"
  :data="{ closeOnBackdropClick: false }"
>
  <div class="vertical-container">
    <div>
      <KButton @click="modal13Visible = true">Modal</KButton>
    </div>
    <KInputSwitch
      v-model="data.closeOnBackdropClick"
      label="Close on backdrop click"
    />
  </div>

  <KModal
    :close-on-backdrop-click="data.closeOnBackdropClick"
    hide-close-icon
    :visible="modal13Visible"
    @cancel="closeAllModals"
    @proceed="closeAllModals"
  >
    Try clicking on modal backdrop.
  </KModal>
</KComponent>

```html
<KModal
  :close-on-backdrop-click="false"
  hide-close-icon
  :visible="modalVisible"
  @cancel="handleModalClose"
  @proceed="handleModalProceed"
>
  Try clicking on modal backdrop.
</KModal>
```

### width

With of the modal window. Defaults to `500px`.

<KButton @click="modal14Visible = true">Modal</KButton>
<KModal
  width="90%"
  title="Modal"
  :visible="modal14Visible"
  @cancel="closeAllModals"
  @proceed="closeAllModals"
/>

```html
<KModal
  width="90%"
  title="Modal"
  :visible="modalVisible"
  @cancel="handleModalClose"
  @proceed="handleModalProceed"
/>
```

> ### maxHeight

Reviewer: suggestion: move width and max-height above the action button props since these are likely more widely used

</KModal>

```html
<KModal
  max-height="200px"
  title="Modal"
  :visible="modalVisible"
  @cancel="handleModalClose"
  @proceed="handleModalProceed"
>
  <p>Lorem ipsum dolor sit amet...
</KModal>
```

### tabbableOptions

Options to be passed to [`focus-trap`](https://github.com/focus-trap/focus-trap), which is responsible for trapping focus inside the modal box.

## Slots

### default

> Slot for modal content.

Reviewer: Can you mention something here about the differences between `default` and `content` slots?

  >
    <template #footer-actions>
      <KButton
        appearance="tertiary"
        @click="handleModalClose">
        <BackIcon />
        Back
      </KButton>
      <KButton @click="handleModalClose">
      Proceed
      <ForwardIcon />
      </KButton>
    </template>
  </KModal>
</template>

<script setup lang="ts">
const modalVisible = ref<boolean>(false)

const handleModalClose = () => {
  modalVisible.value = false
}
</script>
```

:::tip NOTE
KModal takes care of placement and spacing between the buttons when using `footer-actions` slot. However, should you want to mimic that container behaviour using, for example, [`footer`](#footer) slot, we recomend you use `$kui-space-40` [design token](https://github.com/Kong/design-tokens) for spacing between the buttons and place the container containing the buttons on the right using `margin-left: $kui-space-auto;` rule.
:::

### content

By default KModal provides you with the standard layout: modal header (optional), content section and footer. However, should you need to make a custom modal window, you can use `content` slot for your content. It will provide you with just a padded container with white background.

> :::warning NOTE

Reviewer: ```suggestion
:::warning IMPORTANT
```

> Focus-trap requires at least one tabbable element to be present in modal at all times. You can learn more about what elements are considered tabbable [here](https://github.com/focus-trap/tabbable).

Reviewer: Can you provide guidance here on what this means? e.g. ensure your modal has at least one element ... blah

          <KButton @click="modalVisible = false">Show me around</KButton>
        </div>
      </div>
    </template>
  </KModal>
</template>

<script setup lang="ts">
const modalVisible = ref<boolean>(false)
</script>

<style lang="scss" scoped>
:deep(.k-modal .modal-container) {
  padding: $kui-space-0;
}

.modal-content {
  img {
    border-top-left-radius: $kui-border-radius-40;
    border-top-right-radius: $kui-border-radius-40;
  }

  .info-container {
    display: flex;
    flex-direction: column;
    gap: $kui-space-60;
    padding: $kui-space-80;
    text-align: center;

    p {
      color: $kui-color-text-neutral;
    }
  }
}
</style>
```

:::tip NOTE
By default when using `content` slot KModal comes with a padded container. However, should you want to customize that behaviour, you can override that like in the example above. We suggest you use `$kui-space-80` [design token](https://github.com/Kong/design-tokens) token for any container padding and `$kui-border-radius-40` token for rounding the corners.
:::

## Events

### proceed

> Emitted when action button is clicked. Doesn't pass any payload.

Reviewer: ```suggestion
Emitted when action button is clicked.
```
